packages/devtools/src/useReduxDevToolsExtension.tsx

```diff
@@ -0,0 +1,106 @@
+import * as React from 'react';
+import type {
+  NavigationContainerRef,
+  NavigationState,
+  NavigationAction,
+} from '@react-navigation/core';
+import deepEqual from 'deep-equal';
+
+type DevToolsConnection = {
+  init(value: any): void;
+  send(action: any, value: any): void;
+  subscribe(
+    listener: (message: { type: string; [key: string]: any }) => void
+  ): () => void;
+};
+
+type DevToolsExtension = {
+  connect(options: { name: string }): DevToolsConnection;
+  disconnect(): void;
+};
+
+declare const __REDUX_DEVTOOLS_EXTENSION__: DevToolsExtension | undefined;
+
+export default function useReduxDevToolsExtension(
+  ref: React.RefObject<NavigationContainerRef>
+) {
+  const devToolsRef = React.useRef<DevToolsConnection>();
+
+  if (
+    devToolsRef.current === undefined &&
+    typeof __REDUX_DEVTOOLS_EXTENSION__ !== 'undefined'
+  ) {
+    devToolsRef.current = __REDUX_DEVTOOLS_EXTENSION__.connect({
+      name: '@react-navigation/devtools',
+    });
+  }
+
+  const lastStateRef = React.useRef<NavigationState | undefined>();
+  const lastActionRef = React.useRef<NavigationAction | undefined>(undefined);
+
+  React.useEffect(
+    () =>
+      devToolsRef.current?.subscribe((message) => {
+        if (message.type === 'DISPATCH' && message.state) {
+          const state = JSON.parse(message.state);
+
+          lastStateRef.current = state;
+          ref.current?.resetRoot(state);
+        }
+      }),
+    [ref]
+  );
+
+  React.useEffect(() => {
+    const devTools = devToolsRef.current;
+    const navigation = ref.current;
+
+    if (!navigation || !devTools) {
+      return;
+    }
+
+    if (lastStateRef.current === undefined) {
+      const state = navigation.getRootState();
+
+      devTools.init(state);
+      lastStateRef.current = state;
+    }
+
+    const unsubscribeAction = navigation.addListener(
+      '__unsafe_action__',
+      (e) => {
+        const action = e.data.action;
+
+        if (e.data.noop) {
+          // Even if the state didn't change, it's useful to show the action
+          devTools.send(action, lastStateRef.current);
+        } else {
+          lastActionRef.current = action;
+        }
+      }
+    );
+
+    const unsubscribeState = navigation.addListener('state', (e) => {
+      if (lastStateRef.current === e.data.state) {
+        return;
+      }
+
+      const state = navigation.getRootState();
```

Comment:
I'm a little confused about the difference between `e.data.state` and `navigation.getRootState`, and why you do a shallow comparison against the former and a deep comparison against the latter

Reply:
The first check is to skip it if we did the `resetRoot(state)` ourselves, in which case, this state object would be same as that one. Deep equality would work, but a simple referential equality is less expensive.

For using `getRootState()`, it returns fully hydrated state object, so that's what we should be using most of the time anyway.